File: src/com/github/deadleg/idea/openresty/lua/NgxLuaKeywords.java
package com.github.deadleg.idea.openresty.lua;

import org.jetbrains.annotations.Nullable;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

final class NgxLuaKeywords {
    private final static Map<String, List<String>> keywords;
    private final static Map<String, String[]> args;

    static {
        Map<String, List<String>> keywordsTemp = new HashMap<>();
        Map<String, String[]> argsTemp = new HashMap<>();
        try {
            Files.lines(Paths.get(NgxLuaKeywords.class.getClassLoader().getResource("/ngxkeywords.txt").toURI()))
                    .forEach(line -> {
                        String[] content = line.split(" ");
                        String[] fields = content[0].split("\\.");
                        for (int i = 0; i < fields.length; i++) {
                            if (i == 0) {
                                if (!keywordsTemp.containsKey("ngx")) {
                                    keywordsTemp.put("ngx", new ArrayList<>());
                                }
                                keywordsTemp.get("ngx").add(fields[i]);
                            } else {
                                String key = "ngx.";
                                for (int j = 0; j < i; j++) {
                                    key += fields[j] + ".";
                                }
                                key = key.substring(0, key.length() - 1);
                                if (!keywordsTemp.containsKey(key)) {
                                    keywordsTemp.put(key, new ArrayList<>());
                                }
                                keywordsTemp.get(key).add(fields[i]);
                            }
                        }

                        if (content.length > 1) {
                            argsTemp.put(content[0], content[1].split(", "));
                        }
                    });
        } catch (IOException | URISyntaxException e) {
            throw new RuntimeException("Failed to load keyworkds", e);
        }

        keywords = Collections.unmodifiableMap(keywordsTemp);
        args = Collections.unmodifiableMap(argsTemp);
    }

    public static List<String> getKeywords(String key) {
        return keywords.get(key);
    }

    @Nullable
    public static String[] getArgs(String keyword) {
        return args.get(keyword);
    }
}

File: src/com/github/deadleg/idea/openresty/lua/NgxLuaKeywords.kt
package com.github.deadleg.idea.openresty.lua

import java.io.IOException
import java.net.URISyntaxException
import java.nio.file.FileSystems
import java.nio.file.Files
import java.nio.file.Paths
import java.util.*

internal object NgxLuaKeywords {
    /**
     * Contains the key and a list of sub functions e.g. ngx.location = capture
     */
    private val keywords: Map<String, List<String>>
    private val args: Map<String, List<String>>
    /**
     * Set of all keywords e.g. ngx.timer.at, ngx.location.capture
     */
    private val fullKeywords: Set<String>

    init {
        val keywordsTemp = mutableMapOf<String, MutableList<String>>()
        val argsTemp = mutableMapOf<String, MutableList<String>>()
        val fullKeywordsTemp = mutableSetOf<String>()
        try {
            val uri = javaClass.classLoader.getResource("/ngxkeywords.txt")!!.toURI()
            FileSystems.newFileSystem(uri, emptyMap<String, Void>())
            Files.lines(Paths.get(uri)).forEach { line ->
                fullKeywordsTemp.add("ngx." + line.split(' ')[0])
                val content = line.split("\\{?,?\\s\\}?".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                val fields = content[0].split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                for (i in fields.indices) {
                    if (i == 0) {
                        if (!keywordsTemp.containsKey("ngx")) {
                            keywordsTemp.put("ngx", ArrayList())
                        }
                        keywordsTemp["ngx"]!!.add(fields[i])
                    } else {
                        var key = "ngx."
                        for (j in 0 until i) {
                            key += fields[j] + "."
                        }
                        key = key.substring(0, key.length - 1)
                        if (!keywordsTemp.containsKey(key)) {
                            keywordsTemp.put(key, ArrayList())
                        }
                        keywordsTemp[key]!!.add(fields[i])
                    }
                }

                if (content.size > 1) {
                    argsTemp.put("ngx." + content[0], content.drop(1).toMutableList())
                }
            }
        } catch (e: IOException) {
            throw RuntimeException("Failed to load keywords", e)
        } catch (e: URISyntaxException) {
            throw RuntimeException("Failed to load keywokds", e)
        }

        keywords = keywordsTemp
        args = argsTemp
        fullKeywords = fullKeywordsTemp
    }

    fun getKeywords(key: String): List<String>? {
        return keywords[key]
    }

    fun getArgs(keyword: String): List<String>? {
        return args[keyword]
    }

    fun isAKeyword(keyword: String): Boolean {
        return fullKeywords.contains(keyword)
    }
}
